글쓰기: 화면 구현

jwonyLee opened the pull request:
## 개요

<img src="https://user-images.githubusercontent.com/15073405/138447406-95cf835c-c6be-4f42-b39f-ce1d79e1bc6a.png" width="25%" /><img src="https://user-images.githubusercontent.com/15073405/138447411-22d8d95b-d092-4dd9-bad5-a1ae649ff478.png" width="25%" /><img src="https://user-images.githubusercontent.com/15073405/138447402-b4e03ff8-aa94-47f4-9519-0cbeca1b92f4.png" width="25%" /><img src="https://user-images.githubusercontent.com/15073405/138447392-c8999acc-91db-44b1-a409-b77631852efd.png" width="25%" />

글쓰기 화면(샀다, 사고싶다, 선물받았다) 화면 구현

## 작업 사항

- `CenterTextButton` 구현
- `WriteViewController` 구성
  - 전체를 TableView로 구성했습니다. 
- `WriteImageTableCell` 구현
  - 사진 추가, 추가한 사진 썸네일 표시를 위한 TableCell
  - CollectionView를 가지고 있음
- `UIAlertViewController`를 `AlertViewController`로 대체
- 입력 중일 때만 clearButton 표시

## 예외 사항 (미구현)
- 사진 추가, 썸네일 삭제 등 이미지 관련한 로직
- 글 저장 기능
- 카테고리 관련 로직
  - 현재는 인디케이터 버튼을 누르면 #107 과 동일하게 카테고리가 선택되었을 때 모습으로 보여집니다.

### Linked Issue

close #32 
close #128

jwonyLee committed:
[#32] view: CenterTextButton 구현

jwonyLee committed:
[#32] feat: configure data for tableview display

jwonyLee committed:
[#32] view: WriteImageTableCell 구현

jwonyLee committed:
[#32] view: 글쓰기 화면 구성

jwonyLee committed:
[#32] feat: 키보드가 올라왔을 때 스크롤 이동

jwonyLee committed:
[#128] feat: 값이 있고, 입력하는 동안에만 clearButton 노출
- 다른 곳을 터치했을 때 endEditing(true)

jwonyLee committed:
[#32] view: UIAlertViewController를 AlertViewController로 변경

jwonyLee committed:
[#32] view: 누락된 카테고리 항목 추가

jwonyLee committed:
[#32] fix: 피드백 반영
- 텍스트 필드 툴바 변경
- 작성 완료 버튼을 home indicator까지 채우게 변경
- type 일 때 항목 개수 바뀌는 부분 주석으로 명시
- WriteViewModel.Section.free -> .contents 로 이름 변경
- CameraButtonCollectionCell Height 오류 수정

jwonyLee committed:
[#32] fix: 입력 항목 선택 시 스크롤 이동 수정

gustn3965 commented on ThingLog/View/TableVeiwCell/WriteTextViewCell.swift:
```diff
@@ -53,35 +62,39 @@ extension WriteTextViewCell {
+
+        textView.rx.didBeginEditing
+            .bind { [weak self] in
+                self?.isEditingSubject.onNext(true)
```
👀 `Textview`를 `private` 하지 않으면 외부에서 바인딩 한번으로, 이 코드를 줄일 수 있지 않나용?

gustn3965 commented on ThingLog/ViewController/WriteViewController+setup.swift:
```diff
@@ -59,4 +28,85 @@ extension WriteViewController {
+
+        inset = tableView.verticalScrollIndicatorInsets
+        inset.bottom = 0
+        UIView.animate(withDuration: 0.3) {
```
👀 코드가 동일한데, `Height`만 인자로 받아서, 하나의 메서드로 정리하는게 낫지 않을까요?  
그리고 `setupTableView`에서는 `self` 는 클로저 없는 부분에서는 안써도 될 것 같아요

gustn3965 commented on ThingLog/ViewController/WriteViewController.swift:
```diff
@@ -12,6 +12,32 @@ import RxSwift
+        tableView.tableHeaderView = headerLabel
+        tableView.tableFooterView = UIView(frame: CGRect(x: 0, y: 0, width: tableView.frame.size.width, height: 1))
+        if #available(iOS 15.0, *) {
+            tableView.sectionHeaderTopPadding = 0
```
⚠️ 이 코드 중요한 코드 인가요? xcode 12버전에서는` Value of type 'UITableView' has no member 'sectionHeaderTopPadding'`이라면서 에러 떠서 주석처리하고 실행해야 돼서요

jwonyLee replied:
해당 부분이 없으면 iOS 15 환경에서 가장 마지막 셀 하단에 라인이 생겨서 필요한 부분입니다.
![스크린샷 2021-10-25 오후 5 46 50](https://user-images.githubusercontent.com/15073405/138664316-2c05d5dc-41bf-497b-bc3d-f96e6b2af6d9.png)

gustn3965 commented on ThingLog/ViewController/WriteViewController.swift:
```diff
@@ -22,31 +48,137 @@ final class WriteViewController: UIViewController {
+            alertController.dismiss(animated: false) {
+                self?.navigationController?.dismiss(animated: true, completion: {
+                    self?.navigationController?.viewControllers.removeAll()
+                })
```
👀저희가 `Coordinator`를 이용하고 있으니, `navigation` 관련된 로직은` coordinator`에게 전담하는 건 어떨까요?

gustn3965 commented on ThingLog/ViewController/WriteViewController.swift:
```diff
@@ -22,31 +48,137 @@ final class WriteViewController: UIViewController {
+
+// MARK: - Delegate
+extension WriteViewController: WriteTextViewCellDelegate {
+    func updateTextViewHeight(_ cell: WriteTextViewCell, _ textView: UITextView) {
```
👀인자로 받는 `cell`은 사용하고 있지 않는데, 추후에 사용하나요?

gustn3965 commented on ThingLog/ViewController/WriteViewController.swift:
```diff
@@ -12,6 +12,32 @@ import RxSwift
+
     // MARK: - Properties
     var coordinator: Coordinator?
     var viewModel: WriteViewModel?
```
제가 생각하기에는 이 뷰컨트롤러는 `viewModel`이 굉장히 중요하다라고 생각하는데, 맞나요? 
 중요하다고 생각하는 이유는 타입에 따라 뷰모양이 변경되고, 심지어cell개수를 return하는곳에서도 nil인 경우는 0으로 처리하니까요. 

하지만 이를 옵셔널로 처리하여, 뷰컨트롤러 생성할 때 생성자로 받지 않고, 따로 프로퍼티로 직접 접근하여 타입을 할당하고 있더라구요.
```swift
// WriteCoordinator.swift
let writeViewController: WriteViewController = WriteViewController()
writeViewController.coordinator = self
writeViewController.viewModel = WriteViewModel(writeType: type)
``` 

중요한 프로퍼티인데, 따로 프로퍼티로 두는 만큼 반드시 타입을 지정해야한 다는 문구도 없어서, 추후에 `WriteViewController`를 사용할 때 혼란이 생길 것 같습니다. 반드시 필요한 프로퍼티라면 생성자에서 받는게 더 나을 듯 싶습니다. 아니면 그에 따른 주석이 반드시 필요하다고 생각합니다.

gustn3965 reviewed (commented):
리뷰 남겼습니다. 확인 부탁드려요 ~

jwonyLee committed:
[#32] fix: 코드 리뷰 피드백 반영
- WriteTextCell.textView 접근 지정자 변경
- removeTableViewBottomInset() 삭제
- 네비게이션 관련 로직 WriteCoordinator로 이동
- updateTextViewHeight 파라미터 변경
- WriteViewController 초기화 시 WriteViewModel 주입받을 수 있게 변경
- WriteTextCell 최소 높이 지정, separatorInset = .zero 지정

jwonyLee commented:
피드백 반영해서 올렸습니당. 이야기 나왔던 `WriteTextViewCell`의 최소 높이를 380으로 지정하고, `separatorInset = .zero`로 설정해서 마지막 선이 안 나오게 했습니당. 최소 높이는 병연님이랑 협의된 부분입니다~_~

gustn3965 commented on ThingLog/ViewController/WriteViewController.swift:
```diff
@@ -12,41 +12,180 @@ import RxSwift
+        tableView.tableHeaderView = headerLabel
+        tableView.tableFooterView = UIView(frame: CGRect(x: 0, y: 0, width: tableView.frame.size.width, height: 1))
+        if #available(iOS 15.0, *) {
+            tableView.sectionHeaderTopPadding = 0
```
⚠️ 이 코드는 필요없는 코드 인 것 같으니 삭제해주세요. xcode에서 에러납니다.

gustn3965 reviewed (approved):
셀내에서 separatorInset을 이용할 수 있군요! 하나 배워갑니다. 

코멘트단 부분만 수정하시고 머지하세요

jwonyLee committed:
[#32] fix: iOS 15.0 환경에서 동작하는 코드 주석 처리
